externallib.php

```diff
@@ -0,0 +1,688 @@
+                $loggdinuserhardness[] = round($usershardness, 2);
+            }
+            /* questionpercat */
+            $questionpercatdata = array('labels' => $catname, 'datasets' => array(array(
+                'label'
+                => get_string('questionspercategory', 'gradereport_quizanalytics'),
+                'backgroundColor' => $randomcatcolor, 'data' => $catdata
+            )));
+            $questionpercatopt = array('legend' => array(
+                'display' => false,
+                'position' => 'bottom', 'labels' => array('boxWidth' => 13)
+            ), 'title' => array(
+                'display' => true,
+                'position' => 'bottom', 'text' => get_string('questionspercategory', 'gradereport_quizanalytics')
+            ));
+            /* allusers */
+            arsort($overallhardness);
+            $maxhardnesskeys = array_keys($overallhardness, max($overallhardness));
+
+            foreach ($maxhardnesskeys as $maxhardnesskey) {
+                $previous = $maxhardnesskey;
+                break;
+            }
+            $randomoverallhardnesscolor = array();
+            $overallhardnessdata = array();
+            $catnamedata = array();
+            $catcount = 0;
+            foreach ($overallhardness as $key => $val) {
+                if ($totalwrongattemts[$key] > 0) {
+                    $twentyperpreviouswrongattempt = (($totalwrongattemts[$previous] * 20) / 100);
+                    if ($totalwrongattemts[$key] >= $twentyperpreviouswrongattempt) {
+                        if ($catcount < 10) {
+                            $overallhardnessdata[] = $val;
+                            $catnamedata[] = $catname[$key];
+                            $randomoverallhardnesscolor[] = "#" . random_color();
+                            $catcount++;
+                        }
+                    }
+                }
+                $previous = $key;
+            }
+            $allusersdata = array(
+                'labels' => $catnamedata, 'datasets' => array(
+                    array(
+                        'label' => get_string('hardness', 'gradereport_quizanalytics'),
+                        'backgroundColor' => $randomoverallhardnesscolor,
+                        'data' => $overallhardnessdata
+                    )
+                )
+            );
+            $allusersopt = array('legend' => array(
+                'display' => false,
+                'position' => 'bottom'
+            ), 'title' => array(
+                'display' => false,
+                'position' => 'bottom', 'text' => get_string('hardcatalluser', 'gradereport_quizanalytics')
+            ));
+            /* loggedinuser */
+            arsort($loggdinuserhardness);
+            $maxloggdinuserhardnesskeys = array_keys($loggdinuserhardness, max($loggdinuserhardness));
+
+            foreach ($maxloggdinuserhardnesskeys as $maxloggdinuserhardnesskey) {
```

Comment:
give a specific/understandable name... already explain multiple times.. this is d last time mentioned here